sdk/core/System.ClientModel/src/Options/RequestOptions.cs

```diff
@@ -0,0 +1,104 @@
+
+    public RequestOptions()
+    {
+        CancellationToken = CancellationToken.None;
+        ErrorOptions = ClientErrorBehaviors.Default;
+
+        _addHeaders = new PipelineRequestHeaders();
+    }
+
+    public CancellationToken CancellationToken { get; set; }
+
+    public ClientErrorBehaviors ErrorOptions { get; set; }
+
+    public void AddHeader(string name, string value)
+    {
+        Argument.AssertNotNull(name, nameof(name));
+        Argument.AssertNotNull(value, nameof(value));
+
+        AssertNotFrozen();
+
+        _addHeaders.Add(name, value);
+    }
+
+    public void AddPolicy(PipelinePolicy policy, PipelinePosition position)
+    {
+        Argument.AssertNotNull(policy, nameof(policy));
+
+        AssertNotFrozen();
+
+        switch (position)
+        {
+            case PipelinePosition.PerCall:
+                _perCallPolicies = ClientPipelineOptions.AddPolicy(policy, _perCallPolicies);
+                break;
+            case PipelinePosition.PerTry:
+                _perTryPolicies = ClientPipelineOptions.AddPolicy(policy, _perTryPolicies);
+                break;
+            case PipelinePosition.BeforeTransport:
+                _beforeTransportPolicies = ClientPipelineOptions.AddPolicy(policy, _beforeTransportPolicies);
+                break;
+            default:
+                throw new ArgumentException($"Unexpected value for position: '{position}'.");
+        }
+    }
+
+    // Set options on the message before sending it through the pipeline.
+    internal void Apply(PipelineMessage message)
+    {
+        _frozen = true;
+
+        // Set the cancellation token on the message so pipeline policies
+        // will have access to it as the message flows through the pipeline.
+        // This doesn't affect Azure.Core-based clients because the HttpMessage
+        // cancellation token will be set again in HttpPipeline.Send.
+        message.CancellationToken = CancellationToken;
+
+        // We don't overwrite the classifier on the message if it's already set.
+        // This preserves any values set by the client author, and is also
+        // needed for Azure.Core-based clients so we don't overwrite a default
+        // Azure.Core ResponseClassifier.
+        message.MessageClassifier ??= PipelineMessageClassifier.Default;
+
+        // Copy custom pipeline policies to the message.
+        message.PerCallPolicies = _perCallPolicies;
+        message.PerTryPolicies = _perTryPolicies;
+        message.BeforeTransportPolicies = _beforeTransportPolicies;
+
+        // Add the values of any headers set via AddHeader.
+        foreach (var header in _addHeaders)
+        {
+            message.Request.Headers.Add(header.Key, header.Value);
+        }
+    }
+
+    private void AssertNotFrozen()
+    {
+        if (_frozen)
+        {
+            throw new InvalidOperationException("Cannot make changes to RequestOptions after its first use.");
```

Comment:
Would it be more accurate to say it cannot be changed after the related request is sent?

Reply:
We actually freeze the first time the options are applied, which happens before it is sent - if there's a different way to say this that you think is more helpful, I'm happy to update it.

Reply:
I wouldn't want to say it can't be changed after Send because I think that could lead someone to understand what's happening incorrectly in a way that could lead them to make mistakes.  Let me see if I can reword it a bit, though, to improve the clarity...

Reply:
Going with `throw new InvalidOperationException("Cannot change a RequestOptions instance after it has been passed to a client method.");`, but shout if you think something else would be better!